### PR TITLE
improve(cdn): cdn_domain support certificate config.

### PR DESCRIPTION
**Pull request by Lexsss**

### CODE REVIEW

File: alicloud/service_alicloud_cdn.go [hunk 1]
```diff
@@ -21,17 +21,20 @@ func (c *CdnService) DescribeCdnDomain(id string) (*cdn.GetDomainDetailModel, er
...
+	domain, _ := raw.(*cdn.DescribeCdnDomainDetailResponse)
 	if domain.GetDomainDetailModel.DomainName != id {
 		return nil, WrapErrorf(Error(GetNotFoundMessage("cdn_domain", id)), NotFoundMsg, ProviderERROR)
 	}
 	return &domain.GetDomainDetailModel, nil
 }
 
-func (c *CdnService) DescribeDomainConfig(id string) (*cdn.DomainConfig, error) {
+func (c *CdnService) DescribeCdnDomainConfig(id string) (*cdn.DomainConfig, error) {
 	strs := strings.Split(id, ":")
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_cdn.go]: please use ParseResourceId

File: alicloud/service_alicloud_cdn.go [hunk 2]
```diff
@@ -40,10 +43,13 @@ func (c *CdnService) DescribeDomainConfig(id string) (*cdn.DomainConfig, error)
...
 	})
 	if err != nil {
+		if IsExceptedError(err, InvalidDomainNotFound) {
+			return nil, WrapErrorf(err, NotFoundMsg, AlibabaCloudSdkGoERROR)
+		}
 		return nil, WrapErrorf(err, DefaultErrorMsg, id, request.GetActionName(), AlibabaCloudSdkGoERROR)
 	}
-	configs, _ := raw.(*cdn.DescribeCdnDomainConfigsResponse)
 	addDebug(request.GetActionName(), raw)
+	configs, _ := raw.(*cdn.DescribeCdnDomainConfigsResponse)
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_cdn.go]: please change configs to response

File: alicloud/resource_alicloud_cdn_domain_new.go [hunk 3]
```diff
@@ -80,10 +80,51 @@ func resourceAlicloudCdnDomainNew() *schema.Resource {
...
+							Default:      "on",
+							Optional:     true,
+							ValidateFunc: validateCdnEnable,
+						},
+						"private_key": {
+							Type:      schema.TypeString,
+							Optional:  true,
+							Sensitive: true,
+						},
+						"force_set": {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cdn_domain_new.go]: validate?

File: alicloud/resource_alicloud_cdn_domain_new.go [hunk 4]
```diff
@@ -200,43 +257,141 @@ func resourceAlicloudCdnDomainReadNew(d *schema.ResourceData, meta interface{})
...
+	certInfo, err := cdnService.DescribeDomainCertificateInfo(d.Id())
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	if v, ok := d.GetOk("certificate_config"); ok {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cdn_domain_new.go]: computed

File: alicloud/resource_alicloud_cdn_domain_new.go [hunk 5]
```diff
@@ -200,43 +257,141 @@ func resourceAlicloudCdnDomainReadNew(d *schema.ResourceData, meta interface{})
...
-	d.Set("cdn_type", domain.CdnType)
-	d.Set("scope", domain.Scope)
+	d.Set("domain_name", object.DomainName)
+	d.Set("cdn_type", object.CdnType)
+	d.Set("scope", object.Scope)
+
+	certInfo, err := cdnService.DescribeDomainCertificateInfo(d.Id())
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cdn_domain_new.go]: why

File: alicloud/resource_alicloud_cdn_domain_new.go [hunk 6]
```diff
@@ -200,43 +257,141 @@ func resourceAlicloudCdnDomainReadNew(d *schema.ResourceData, meta interface{})
...
+	}
+	if v, ok := val["cert_type"]; ok && v.(string) != "" {
+		request.CertType = v.(string)
+	}
+
+	if okServerCertificate {
+		request.ServerCertificate = serverCertificate.(string)
+	}
+	if okPrivateKey {
+		request.PrivateKey = privateKey.(string)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cdn_domain_new.go]: merge with 351

File: alicloud/resource_alicloud_cdn_domain_new.go [hunk 7]
```diff
@@ -200,43 +257,141 @@ func resourceAlicloudCdnDomainReadNew(d *schema.ResourceData, meta interface{})
...
+	}
 
-			return resource.NonRetryableError(WrapError(err))
+	if request.ServerCertificateStatus == "off" {
+		if oldConfig == nil || len(oldConfig) == 0 {
+			if okServerCertificate || okPrivateKey {
+				return WrapError(Error("If 'server_certificate_status' value is 'off', you can not set the value of 'server_certificate' and 'private_key'."))
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cdn_domain_new.go]: diffsuppress

File: alicloud/resource_alicloud_cdn_domain_new.go [hunk 8]
```diff
@@ -200,43 +257,141 @@ func resourceAlicloudCdnDomainReadNew(d *schema.ResourceData, meta interface{})
...
+			}
+		}
+	} else {
+		if !okServerCertificate || !okPrivateKey {
+			return WrapError(Error("If 'server_certificate_status' value is 'on', you must set 'server_certificate', and 'private_key'"))
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cdn_domain_new.go]: removed

File: alicloud/service_alicloud_cdn.go [hunk 9]
```diff
@@ -12,7 +12,7 @@ type CdnService struct {
 	client *connectivity.AliyunClient
 }
 
-func (c *CdnService) DescribeCdnDomain(id string) (*cdn.GetDomainDetailModel, error) {
+func (c *CdnService) DescribeCdnDomainNew(id string) (*cdn.GetDomainDetailModel, error) {
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_cdn.go]: check with old resource

File: alicloud/validators.go [hunk 10]
```diff
@@ -1132,6 +1132,16 @@ func validateCdnEnable(v interface{}, k string) (ws []string, errors []error) {
 	return
 }
 
+func validateCdnCert(v interface{}, k string) (ws []string, errors []error) {
```
Comment (xiaozhu36) [inline alicloud/validators.go]: Using `validation.StringInSlice()` instead.

File: website/docs/r/cdn_domain_new.html.markdown [hunk 11]
```diff
@@ -42,6 +42,7 @@ The following arguments are supported:
 * `cdn_type` - (Required, ForceNew) Cdn type of the accelerated domain. Valid values are `web`, `download`, `video`.
 * `scope` - (Optional) Scope of the accelerated domain. Valid values are `domestic`, `overseas`, `global`. Default value is `domestic`. This parameter's setting is valid Only for the international users and domestic L3 and above users .
 * `sources` - (Optional, Type: list) The source address list of the accelerated domain. Defaults to null. See Block Sources.
+* `certificate_config` - (Optional, Type: list, Available in 1.51.0+)  Certificate config of the accelerated domain. It's a list and consist of at most 1 item.
```
Comment (xiaozhu36) [inline website/docs/r/cdn_domain_new.html.markdown]: 1.52.0

File: website/docs/r/cdn_domain_new.html.markdown [hunk 12]
```diff
@@ -53,6 +54,17 @@ The `sources` block supports the following:
...
+### Block certificate_config
+
+The `certificate_config` block supports the following:
+
+* `server_certificate_status` - (Optional, Available in 1.51.0+) This parameter indicates whether or not enable https. Valid values are `on` and `off`. Default value is `on`.
+* `server_certificate` - (Optional, Available in 1.51.0+) The SSL server certificate string. This is required if `server_certificate_status` is `on`
+* `private_key` - (Optional, Available in 1.51.0+) The SSL private key. This is required if `server_certificate_status` is `on`
+* `force_set` - (Optional, Available in 1.51.0+) When you set to 1, ignore the repeated verification for certificate name, and cover the information of the origin certificate (with the same name).
+* `cert_name` - (Optional, Available in 1.51.0+) The SSL certificate name.
+* `cert_type` - (Optional, Available in 1.51.0+) The SSL certificate type, can be "upload", "cas" and "free".
```
Comment (xiaozhu36) [inline website/docs/r/cdn_domain_new.html.markdown]: remove available

File: alicloud/resource_alicloud_cdn_domain_new.go [hunk 13]
```diff
@@ -80,10 +80,51 @@ func resourceAlicloudCdnDomainNew() *schema.Resource {
...
+						},
+						"private_key": {
+							Type:      schema.TypeString,
+							Optional:  true,
+							Sensitive: true,
+						},
+						"force_set": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validateAllowedStringValue([]string{"1", "0"}),
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cdn_domain_new.go]: Using validation.StringInSlice() instead.

File: alicloud/resource_alicloud_cdn_domain_new.go [hunk 14]
```diff
@@ -200,43 +257,126 @@ func resourceAlicloudCdnDomainReadNew(d *schema.ResourceData, meta interface{})
...
+
+	certInfo, err := cdnService.DescribeDomainCertificateInfo(d.Id())
+	if err != nil {
+		if NotFoundError(err) {
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	oldConfig := d.Get("certificate_config").([]interface{})
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cdn_domain_new.go]: if the certificate_config has default, please it to `Computed: true`

**Pull request merged**